### PR TITLE
KAFKA-14936: Check the versioned table's history retention and compare to grace period (4/N)

### PR DESCRIPTION
**Pull request by wcarlson5**
Check the history retention of the ktable of the grace period join. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java [hunk 1]
```diff
@@ -164,6 +165,53 @@ public void shouldFailIfTableIsNotVersioned() {
...
+
+        final IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, builder::build);
+        assertThat(
+            exception.getMessage(),
+            is("KTable must be versioned to use a grace period in a stream table join.")
+        );
+    }
+
+    @Test
+    public void shouldNotFailIfTableIsVersionedButMaterializationIsInherited() {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:188]: Do we also need a test where the materialization is not inherited?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamTableJoinNode.java [hunk 2]
```diff
@@ -64,6 +70,16 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
         // Steam - KTable join only
         if (otherJoinSideNodeName != null) {
             topologyBuilder.connectProcessorAndStateStores(processorName, storeNames);
+            if (gracePeriod != null) {
+                for (final String storeName : storeNames) {
+                    if (!topologyBuilder.isStoreVersioned(storeName)) {
+                        throw new IllegalArgumentException("KTable must be versioned to use a grace period in a stream table join.");
+                    }
+                    if (gracePeriod.toMillis() > topologyBuilder.getHistoryRetention(storeName)) {
+                        throw new IllegalArgumentException("History history retention must be at least grace period, but it should be larger.");
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamTableJoinNode.java]:
```suggestion
                        throw new IllegalArgumentException("History retention must be at least grace period.");
```
Must history retention be larger than or is it enough to be equal to the grace period? That is not clear from your original error message.

**Review by cadonna — COMMENTED** [review]
@wcarlson5 Thanks for the PR!

Here my feedback!

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java (hunk 1 above)
Comment (vcrfxia) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:188]: I believe those already exist in this file -- there's `shouldFailIfTableIsNotVersioned()` to check the negative case and `shouldDelayJoinByGracePeriod()` for the positive case -- so I think we should be good?

A minor suggestion: a bunch of these new tests look similar even though they test different setups and to an unfamiliar reader it's difficult to pick out which lines of the test are important. You could add a comment to the `filter()` line in the topology to call out that the materialization as a versioned store is inherited through this node; that's the crux of this test that distinguishes it from the other test which already exists.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java [hunk 3]
```diff
@@ -164,6 +165,53 @@ public void shouldFailIfTableIsNotVersioned() {
...
+            Materialized.as(Stores.persistentVersionedKeyValueStore("tableB", Duration.ofMinutes(5))));
+        final KTable<String, String> tableB = source.filter((k, v) -> true);
+        streamA.join(tableB, (value1, value2) -> value1 + value2, Joined.with(Serdes.String(), Serdes.String(), Serdes.String(), "first-join", Duration.ofMillis(6))).to("out-one");
+
+        //should not throw an error
+        builder.build();
+    }
+
+    @Test
+    public void shouldFailIfGracePeriodIsLongerThanHistoryRetention() {
```
Comment (vcrfxia) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:205]: Can you also add the equivalent of this test, for the case where materialization as a versioned store is inherited?

**Review by vcrfxia — COMMENTED** [review]
Thanks @wcarlson5 ! One request for an additional test, but otherwise LGTM.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java (hunk 1 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:188]: Ah, I see! Thanks for the clarification!

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamTableJoinNode.java (hunk 2 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamTableJoinNode.java]: This change is fine, just equal is enough, if not recommended

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java (hunk 3 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:205]: Sure, easy enough

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java (hunk 1 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:188]: good idea

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamTableJoinNode.java [hunk 4]
```diff
@@ -31,17 +32,22 @@ public class StreamTableJoinNode<K, V> extends GraphNode {
...
+                               final String otherJoinSideNodeName,
+                               final Duration gracePeriod) {
         super(nodeName);
 
         // in the case of Stream-Table join the state stores associated with the KTable
         this.storeNames = storeNames;
         this.processorParameters = processorParameters;
         this.otherJoinSideNodeName = otherJoinSideNodeName;
+        this.gracePeriod = gracePeriod;
+
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/StreamTableJoinNode.java]: nit:
```suggestion
```

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBytesStoreTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    public void shouldCreateWriteBatches() {
+        final String key = "a";
+        final Collection<ConsumerRecord<byte[], byte[]>> records = new ArrayList<>();
+        records.add(new ConsumerRecord<>("", 0, 0L, serializeKey(key, 0, 0L).get(), serializeValue(50L)));
+        records.add(new ConsumerRecord<>("", 0, 0L, serializeKey(key, 1, 1L).get(), serializeValue(100L)));
+        final Map<KeyValueSegment, WriteBatch> writeBatchMap = bytesStore.getWriteBatches(records);
+        assertEquals(1, writeBatchMap.size());
+
+        for (final WriteBatch batch : writeBatchMap.values()) {
+            // 2 includes base and index record
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBytesStoreTest.java]: This comment lies. There is no index record here. Please remove.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBytesStoreTest.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        bytesStore.init((StateStoreContext) context, bytesStore);
+    }
+
+    @AfterEach
+    public void close() {
+        bytesStore.close();
+    }
+
+    @Test
+    public void shouldCreateWriteBatches() {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBytesStoreTest.java]: Could you please add a test for the edge case when the the collection of records is empty?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java [hunk 7]
```diff
@@ -164,6 +164,75 @@ public void shouldFailIfTableIsNotVersioned() {
...
+        final StreamsBuilder builder = new StreamsBuilder();
+        final Properties props = new Properties();
+        props.put(StreamsConfig.TOPOLOGY_OPTIMIZATION_CONFIG, StreamsConfig.NO_OPTIMIZATION);
+        final KStream<String, String> streamA = builder.stream("topic", Consumed.with(Serdes.String(), Serdes.String()));
+        final KTable<String, String> tableB = builder.table("topic2", Consumed.with(Serdes.String(), Serdes.String()),
+            Materialized.as(Stores.persistentVersionedKeyValueStore("tableB", Duration.ofMinutes(5))));
+
+        streamA.join(tableB, (value1, value2) -> value1 + value2, Joined.with(Serdes.String(), Serdes.String(), Serdes.String(), "first-join", Duration.ofMinutes(6))).to("out-one");
+        final IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> builder.build(props));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:215]: Could you please structure your tests in such a way that it is clearer what the setup, the actual call under test and the verification is. I usually separate these three parts by an empty line. A clear separation is not always possible, though.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java [hunk 8]
```diff
@@ -164,6 +164,75 @@ public void shouldFailIfTableIsNotVersioned() {
...
+        props.put(StreamsConfig.TOPOLOGY_OPTIMIZATION_CONFIG, StreamsConfig.NO_OPTIMIZATION);
+        final KStream<String, String> streamA = builder.stream("topic", Consumed.with(Serdes.String(), Serdes.String()));
+        final KTable<String, String> source = builder.table("topic2", Consumed.with(Serdes.String(), Serdes.String()),
+            Materialized.as(Stores.persistentVersionedKeyValueStore("V-grace", Duration.ofMinutes(0))));
+        final KTable<String, String> tableB = source.filter((k, v) -> true);
+        // the filter operation forces the table materialization to be inherited
+
+        streamA.join(tableB, (value1, value2) -> value1 + value2, Joined.with(Serdes.String(), Serdes.String(), Serdes.String(), "first-join", Duration.ofMillis(6))).to("out-one");
+
+        final IllegalArgumentException exception = assertThrows(IllegalArgumentException.class, () -> builder.build(props));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:232]: I was wondering if it would not be better testing this exceptions in `StreamTableJoinNodeTest` (does not yet exist). In such a way, the tests might become simpler, because we could verify the exceptions just for the case that a versioned state store is connected to the node in `StreamTableJoinNodeTest`. Here we would merely test if the materializations are correctly inherited (maybe there are already such tests somewhere).
You do not need to do that for this PR. 
But we should try to avoid writing more tests than necessary and at the same time we should cover all cases.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBytesStore.java [hunk 9]
```diff
@@ -42,7 +46,29 @@ protected KeyValue<Bytes, byte[]> getIndexKeyValue(final Bytes baseKey, final by
 
     @Override
     Map<KeyValueSegment, WriteBatch> getWriteBatches(final Collection<ConsumerRecord<byte[], byte[]>> records) {
-        throw new UnsupportedOperationException("Do not use for TimeOrderedKeyValueStore");
+        final Map<KeyValueSegment, WriteBatch> writeBatchMap = new HashMap<>();
```
Comment (vcrfxia) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBytesStore.java]: Looks like this PR is now also adding restoration logic, in addition to comparing table history retention and join grace period. Was that intentional? Whenever we add the restoration logic, we'll want to add integration tests for that -- maybe it'd be easier to have it be a separate PR. WDYT?

**Review by vcrfxia — COMMENTED** [review]
Thanks @wcarlson5 . The changes in this PR for checking table history retention against join grace period LGTM. I notice that the latest commit also introduces restoration functionality, for which we'll want additional test coverage. Should that be split out into a separate PR? Can also add the integration tests into this PR and merge both pieces of functionality together, but that'd require more review iterations on this PR -- up to you.

**Review by cadonna — COMMENTED** [review]
Thanks for the updates, @wcarlson5 !

Here my feedback!

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBytesStore.java (hunk 9 above)
Comment (wcarlson5) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBytesStore.java]: Yeah that should be in a different PR, not sure how that snuck in

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBytesStoreTest.java (hunk 6 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBytesStoreTest.java]: Gonna move this file to a different PR, but also yes

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java (hunk 7 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:215]: I added an empty line I hope that helps

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java (hunk 8 above)
Comment (wcarlson5) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamKTableJoinTest.java:232]: I think there is. For some reason it seems that streams has some aversion to making tests like `StreamTableJoinNodeTest` as not are set up like this

**Review by vcrfxia — APPROVED** [review]
Thanks @wcarlson5 ! LGTM

**Review by cadonna — APPROVED** [review]
Thanks @wcarlson5 !

LGTM!

**Comment by cadonna** [comment]
Build failures are unrelated:
```
    Build / JDK 20 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()
    Build / JDK 20 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()
    Build / JDK 20 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testReplication()
    Build / JDK 20 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.IdentityReplicationIntegrationTest.testReplicateSourceDefault()
    Build / JDK 17 and Scala 2.13 / integration.kafka.server.FetchFromFollowerIntegrationTest.testRackAwareRangeAssignor()
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testBalancePartitionLeaders()
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationExactlyOnceTest.testOffsetTranslationBehindReplicationFlow()
```
